sys/usb/usbus/usbus.c

```diff
@@ -0,0 +1,667 @@
+        return 1;
+    }
+    return 0;
+}
+
+usbus_string_t *_get_descriptor(usbus_t *usbus, uint16_t idx)
+{
+    for (usbus_string_t * str = usbus->strings; str; str = str->next) {
+        if (str->idx == idx) {
+            return str;
+        }
+    }
+    return NULL;
+}
+
+uint16_t usbus_add_interface(usbus_t *usbus, usbus_interface_t *iface)
+{
+    mutex_lock(&usbus->lock);
+    iface->next = usbus->iface;
+    usbus->iface = iface;
+    mutex_unlock(&usbus->lock);
+    return iface->idx;
+}
+
+void usbus_register_event_handler(usbus_t *usbus, usbus_handler_t *handler)
+{
+    mutex_lock(&usbus->lock);
+    handler->next = usbus->handler;
+    usbus->handler = handler;
+    mutex_unlock(&usbus->lock);
+    handler->driver->init(usbus, handler);
+}
+
+int usbus_add_endpoint(usbus_t *usbus, usbus_interface_t *iface, usbus_endpoint_t* ep, usb_ep_type_t type, usb_ep_dir_t dir, size_t len)
+{
+    int res = -ENOMEM;
+    mutex_lock(&usbus->lock);
+    usbdev_ep_t* usbdev_ep = usbus->dev->driver->new_ep(usbus->dev, type, dir, len);
+    if (usbdev_ep) {
+        ep->maxpacketsize = usbdev_ep->len;
+        usbdev_ep->context = usbus;
+        usbdev_ep->cb = _event_ep_cb;
+        ep->ep = usbdev_ep;
+        ep->next = iface->ep;
+        iface->ep = ep;
+        res = 0;
+    }
+    mutex_unlock(&usbus->lock);
+    return res;
+}
+
+static void _activate_endpoints(usbus_t *usbus)
+{
+    for (usbus_interface_t *iface = usbus->iface; iface; iface = iface->next) {
+        for (usbus_endpoint_t *ep = iface->ep; ep; ep = ep->next) {
+            if (ep->active) {
+                static const usbopt_enable_t enable = USBOPT_ENABLE;
+                ep->ep->driver->set(ep->ep, USBOPT_EP_ENABLE, &enable, sizeof(usbopt_enable_t));
+                printf("activated endpoint %d, dir %s\n", ep->ep->num, ep->ep->dir == USB_EP_DIR_OUT? "out" : "in");
+            }
+        }
+        for (usbus_interface_alt_t *alt = iface->alts; alt; alt = alt->next) {
+            for (usbus_endpoint_t *ep = alt->ep; ep; ep = ep->next) {
+                if (ep->active) {
+                    static const usbopt_enable_t enable = USBOPT_ENABLE;
+                    ep->ep->driver->set(ep->ep, USBOPT_EP_ENABLE, &enable, sizeof(usbopt_enable_t));
+                    printf("activated endpoint %d, dir %s\n", ep->ep->num, ep->ep->dir == USB_EP_DIR_OUT? "out" : "in");
+                }
+            }
+        }
+    }
+}
+
+static void _usbus_config_ep0(usbus_t *usbus)
+{
+    static const usbopt_enable_t enable = USBOPT_ENABLE;
+    usbus->in->driver->set(usbus->in, USBOPT_EP_ENABLE, &enable, sizeof(usbopt_enable_t));
+    usbus->out->driver->set(usbus->out, USBOPT_EP_ENABLE, &enable, sizeof(usbopt_enable_t));
+    usbus->out->driver->ready(usbus->out, 0);
+}
+
+void _req_status(usbus_t *usbus)
+{
+    uint8_t status[2];
+    memset(status, 0, 2);
+    usbus_put_bytes(usbus, status, sizeof(status));
+    usbus->in->driver->ready(usbus->in, 2);
```

Comment:
Consider `2` => `sizeof(status)` if that makes sense

Reply:
Done